client/src/main/scala/com.crobox.clickhouse/stream/ClickhouseSink.scala

```diff
@@ -29,24 +44,55 @@ object ClickhouseSink extends LazyLogging {
           else None
       )
       .getOrElse(indexerGeneralConfig)
-    val batchSize     = mergedIndexerConfig.getInt("batch-size")
+    val batchSize = mergedIndexerConfig.getInt("batch-size")
     val flushInterval = mergedIndexerConfig.getDuration("flush-interval").getSeconds.seconds
-    Flow[Insert]
+    Flow[TableOperation]
       .groupBy(Int.MaxValue, _.table)
       .groupedWithin(batchSize, flushInterval)
-      .mapAsyncUnordered(mergedIndexerConfig.getInt("concurrent-requests"))(inserts => {
-        val table       = inserts.head.table
-        val insertQuery = s"INSERT INTO $table FORMAT JSONEachRow"
-        val payload     = inserts.map(_.jsonRow)
-        logger.debug(s"Inserting ${inserts.size} entries in table: $table. Group Within: ($batchSize - $flushInterval)")
-        client
-          .execute(insertQuery, payload.mkString("\n"))
-          .recover {
-            case ex => throw ClickhouseIndexingException("failed to index", ex, payload, table)
-          }
-          .map(_ => inserts)
+      .mapAsync(mergedIndexerConfig.getInt("concurrent-requests"))(operations => {
+        val table = operations.head.table
+        var insertPayload: Seq[String] = Seq.empty
+        var optimizeQuery: Option[String] = None
+
+        operations.foreach {
```

Comment:
This setup is sub optimal. Please use a .collect or other way to 'split' the stream/batch of operation into substreams belonging to each type. Fine to use a pattern match 
```
.map {
  case op:InsertOperation => ..
case op:OptimizeOperation => ...
```

Please find the best / most optimized solution